SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialComponentTest/SpatialComponentTest.cpp

```diff
@@ -0,0 +1,231 @@
+					FinishStep();
+				});
+
+		AddStep(
+			TEXT("Replicated Dynamic Actor Spawned On Same Server - Verify components"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
+			[this](float DeltaTime) {
+				CheckComponents(DynamicReplicatedActor, 1, 2, 1);
+			},
+			5.0f);
+
+		AddStepFromDefinition(ReplicatedDestroyStepDefinition, FWorkerDefinition::Server(1));
+	}
+
+	// Replicated Dynamic Actor Spawned On Different Server. Server 1 should have Authority on BeginPlay, Server 2 on Tick
+	{
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Different Server - Spawn"), FWorkerDefinition::Server(1), nullptr, [this]() {
+			ASpatialComponentTestReplicatedActor* Actor =
+				GetWorld()->SpawnActor<ASpatialComponentTestReplicatedActor>(Server2Position, FRotator::ZeroRotator);
+			CrossServerSetDynamicReplicatedActor(Actor);
+			FinishStep();
+		});
+
+		AddStep(
+			TEXT("Replicated Dynamic Actor Spawned On Different Server - Verify components "
+				 "on Tick"),
+			FWorkerDefinition::AllWorkers, nullptr, nullptr,
+			[this](float DeltaTime) {
+				CheckComponentsCrossServer(DynamicReplicatedActor, 1, 2);
+			},
+			5.0f);
+
+		AddStepFromDefinition(ReplicatedDestroyStepDefinition, FWorkerDefinition::Server(2));
+	}
+}
+
+void ASpatialComponentTest::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(ASpatialComponentTest, DynamicReplicatedActor);
+}
+
+// Checks the number of components on the servers and clients when an actor does not migrate
+void ASpatialComponentTest::CheckComponents(ASpatialComponentTestActor* Actor, int ExpectedServerId, int ExpectedClient1ComponentCount,
+											int ExpectedClient2ComponentCount)
+{
+	const FWorkerDefinition& LocalWorkerDefinition = GetLocalFlowController()->WorkerDefinition;
+	if (LocalWorkerDefinition.Type == ESpatialFunctionalTestWorkerType::Server)
+	{
+		// Allow it to continue working in Native / Single worker setups.
+		if (GetNumberOfServerWorkers() > 1)
+		{
+			if (LocalWorkerDefinition.Id == ExpectedServerId)
+			{
+				RequireTrue(VerifyTestActorComponents(Actor, 2), "Server auth - OnAuthorityGained component and OnActorReady component");
+				FinishStep();
+			}
+			else if (Actor->bNetStartup)
+			{
+				RequireTrue(VerifyTestActorComponents(Actor, 1),
+							"Non-auth servers - Level actors receive OnActorReady only OnAuthorityGained");
+				FinishStep();
+			}
+			else
+			{
+				RequireTrue(VerifyTestActorComponents(Actor, 0),
+							"Non-auth servers - Dynamic actors do not receive OnActorReady or OnAuthorityGained ");
+				FinishStep();
+			}
+		}
+		else // Support for Native / Single Worker.
+		{
+			RequireTrue(VerifyTestActorComponents(Actor, 2),
+						"Native / Single Worker - OnActorReady component and OnAuthorityGained component");
+			FinishStep();
+		}
+	}
+	else // Clients
+	{
+		if (LocalWorkerDefinition.Id == 1)
```

Comment:
Why are these counts different? Could you add a comment to the code please.

Reply:
Depends on which client has gained / lost ownership. The function here depends on the arguments passed so I cannot really comment inside the function but I have added comments to the calling steps. However, on the follow-on task from this one I have split this logic and moved it back to the step so it will be clearer.